crates/goose/src/agents/skills_extension.rs

```diff
@@ -0,0 +1,470 @@
+            supporting_files,
+        })
+    }
+
+    fn parse_frontmatter(content: &str) -> Result<(SkillMetadata, String)> {
+        let lines: Vec<&str> = content.lines().collect();
+
+        if lines.is_empty() || !lines[0].trim().starts_with("---") {
+            return Err(anyhow::anyhow!("Missing YAML frontmatter"));
+        }
+
+        let mut end_index = None;
+        for (i, line) in lines.iter().enumerate().skip(1) {
+            if line.trim().starts_with("---") {
+                end_index = Some(i);
+                break;
+            }
+        }
+
+        let end_index = end_index.ok_or_else(|| anyhow::anyhow!("Unclosed YAML frontmatter"))?;
+
+        let yaml_content = lines[1..end_index].join("\n");
+        let metadata: SkillMetadata = serde_yaml::from_str(&yaml_content)?;
+
+        let body = lines[end_index + 1..].join("\n").trim().to_string();
+
+        Ok((metadata, body))
+    }
+
+    fn find_supporting_files(directory: &Path, skill_file: &Path) -> Result<Vec<PathBuf>> {
+        let mut files = Vec::new();
+
+        if let Ok(entries) = std::fs::read_dir(directory) {
+            for entry in entries.flatten() {
+                let path = entry.path();
+                if path.is_file() && path != skill_file {
+                    files.push(path);
+                } else if path.is_dir() {
+                    if let Ok(sub_entries) = std::fs::read_dir(&path) {
+                        for sub_entry in sub_entries.flatten() {
+                            let sub_path = sub_entry.path();
+                            if sub_path.is_file() {
+                                files.push(sub_path);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        Ok(files)
+    }
+
+    fn discover_skills(&self) -> HashMap<String, Skill> {
+        let mut skills = HashMap::new();
+
+        for dir in self.get_skill_directories() {
+            if let Ok(entries) = std::fs::read_dir(&dir) {
+                for entry in entries.flatten() {
+                    let path = entry.path();
+                    if path.is_dir() {
+                        let skill_file = path.join("SKILL.md");
+                        if skill_file.exists() {
+                            if let Ok(skill) = Self::parse_skill_file(&skill_file) {
+                                skills.insert(skill.metadata.name.clone(), skill);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        skills
+    }
+
+    fn generate_instructions(&self) -> String {
+        let skills = self.discover_skills();
+
+        if skills.is_empty() {
+            return "No skills available.".to_string();
+        }
+
+        let mut instructions = String::from("You have these skills at your disposal, when it is clear they can help you solve a problem or you are asked to use them:\n\n");
```

Comment:
note for others: this is the important part, as the instructions give the skill names. This hints that names would be important, and also how skills are progressively loaded (via `loadSkill`)